Userland/Libraries/LibWebSocket/WebSocket.cpp

```diff
@@ -0,0 +1,597 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Base64.h>
+#include <AK/Debug.h>
+#include <AK/Random.h>
+#include <LibCrypto/Hash/HashManager.h>
+#include <LibWebSocket/Impl/TCPWebSocketConnectionImpl.h>
+#include <LibWebSocket/Impl/TLSv12WebSocketConnectionImpl.h>
+#include <LibWebSocket/WebSocket.h>
+#include <stdio.h>
+#include <unistd.h>
+
+namespace WebSocket {
+
+// Note : The websocket protocol is defined by RFC 6455, found at https://tools.ietf.org/html/rfc6455
+// In this file, section numbers will refer to the RFC 6455
+
+NonnullRefPtr<WebSocket> WebSocket::create(ConnectionInfo connection)
+{
+    return adopt(*new WebSocket(connection));
+}
+
+WebSocket::WebSocket(ConnectionInfo connection)
+    : m_connection(connection)
+{
+}
+
+WebSocket::~WebSocket()
+{
+}
+
+void WebSocket::start()
+{
+    VERIFY(!m_impl);
+    if (m_connection.is_secure())
+        m_impl = TLSv12WebSocketConnectionImpl::construct();
+    else
+        m_impl = TCPWebSocketConnectionImpl::construct();
+
+    m_impl->on_connection_error = [this] {
+        dbgln("WebSocket: Connection error (underlying socket)");
+        fatal_error(WebSocket::Error::CouldNotEstablishConnection);
+    };
+    m_impl->on_connected = [this] {
+        if (m_state != WebSocket::InternalState::EstablishingProtocolConnection)
+            return;
+        m_state = WebSocket::InternalState::SendingClientHandshake;
+        send_client_handshake();
+        drain_read();
+    };
+    m_impl->on_ready_to_read = [this] {
+        drain_read();
+    };
+    m_impl->connect(m_connection);
+}
+
+WebSocket::ReadyState WebSocket::ready_state()
+{
+    switch (m_state) {
+    case WebSocket::InternalState::EstablishingProtocolConnection:
+        return WebSocket::ReadyState::Connecting;
+    case WebSocket::InternalState::SendingClientHandshake:
+        return WebSocket::ReadyState::Connecting;
+    case WebSocket::InternalState::WaitingForServerHandshake:
+        return WebSocket::ReadyState::Connecting;
+    case WebSocket::InternalState::Open:
+        return WebSocket::ReadyState::Open;
+    case WebSocket::InternalState::Closing:
+        return WebSocket::ReadyState::Closing;
+    case WebSocket::InternalState::Closed:
+        return WebSocket::ReadyState::Closed;
+    case WebSocket::InternalState::Errored:
+        return WebSocket::ReadyState::Closed;
+    default:
+        VERIFY_NOT_REACHED();
+        return WebSocket::ReadyState::Closed;
+    }
+}
+
+void WebSocket::send(Message message)
+{
+    if (m_state != WebSocket::InternalState::Open)
+        return;
```

Comment:
Maybe this should `VERIFY()` instead? Calling `send()` on a non-open websocket seems like something that shouldn't happen, and just returning silently seems kinda defensive..

Reply:
That makes sense. Since it might be surprising to clients who are using the `send()` and `close()` methods when these only work on opened sockets, I added some comments explaining that in the `WebSocket.h` header. And I updated the `telws` program to not crash when trying that :)